components/support/rc_crypto/src/hkdf.rs

```diff
@@ -0,0 +1,95 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::{
+    digest,
+    error::*,
+    hmac, p11,
+    util::{ensure_nss_initialized, map_nss_secstatus},
+};
+use nss_sys::*;
+use std::{mem, os::raw::c_uchar, ptr};
+
+pub fn extract_and_expand(
+    salt: &hmac::SigningKey,
+    secret: &[u8],
+    info: &[u8],
+    out: &mut [u8],
+) -> Result<()> {
+    let prk = extract(salt, secret)?;
+    expand(&prk, info, out)?;
+    Ok(())
+}
+
+pub fn extract(salt: &hmac::SigningKey, secret: &[u8]) -> Result<hmac::SigningKey> {
+    let prk = hmac::sign(salt, secret)?;
+    Ok(hmac::SigningKey::new(salt.digest_algorithm(), prk.as_ref()))
+}
+
+pub fn expand(prk: &hmac::SigningKey, info: &[u8], out: &mut [u8]) -> Result<()> {
+    let (mech, block_size) = match prk.digest_algorithm() {
+        digest::Algorithm::SHA256 => (CKM_NSS_HKDF_SHA256, 512 / 8),
+    };
+    ensure_nss_initialized();
+    let hkdf_params = Box::new(CK_NSS_HKDFParams {
+        bExtract: CK_FALSE,
+        pSalt: ptr::null_mut(),
+        ulSaltLen: 0,
+        bExpand: CK_TRUE,
+        pInfo: info.as_ptr() as CK_BYTE_PTR,
+        ulInfoLen: info.len() as u64 as CK_ULONG,
+    });
```

Comment:
It's hard for me to understand exactly what this function does, because `PK11_Derive` is a pretty nondescript name, and it *looks* like you're passing in parameters for a full HKDF, but using a `NULL` salt, so `PK11_Derive` only does the expand step.  It would be great to have some narrative docs here to explain why these various params cause `PK11_Derive` to do the right thing...but IIUC this is mostly copied from the webcrypto implementation?

Perhaps we can at least leave a pointer to the reference code from which this was copied.